### PR TITLE
[crypto] Refactor how the cryptolib AES driver represents configuration parameters.

### PR DESCRIPTION
**Pull request by jadephilipoom**
Previously, the cryptolib's AES driver took an `aes_params_t` struct for all configuration/key/IV information in order to program the hardware. In practice, particularly for AES-GCM, this was a little awkward to work with; if you need to run multiple AES operations with the same key but a different IV, or need to first encrypt and then decrypt, you need to either manipulate the parameters struct or pass around multiple arguments for separate elements such as the key.

Now, I have changed the API to mostly replace `aes_params_t` with `aes_key_t`. The IV is now a separate parameter (since it may well change with the same key; this also allows us to mark it `const` because we don't have to manipulate it within the struct) and the encrypt/decrypt boolean is hidden and replaced with two separate wrappers. This is for readability; now it's easy to see in the code which operation is being performed without having to find `.encrypt = true` wherever the parameters struct was created.

**TL;DR rather than calling `aes_begin(params)` you now call `aes_encrypt_begin(key, iv)`.**

This PR also adds an option to the key struct that configures the block to read a sideloaded key.

### CODE REVIEW

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 1]
```diff
@@ -399,16 +399,20 @@ aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
...
 
 /**
- * One-shot version of the AES API for a single block, for convenience.
+ * One-shot version of the AES encryption API for a single block.
  */
-static aes_error_t aes_single_block(const aes_params_t params,
-                                    const aes_block_t *input,
-                                    aes_block_t *output) {
-  aes_error_t err = aes_begin(params);
+static aes_error_t aes_encrypt_block(const aes_key_t key, const uint32_t iv[4],
```
Comment (moidx) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:404]: re. `const uint32_t iv[4]`. 

It is recommended that for parameters that are exposed in the public interface, to wrap them inside a struct for type safety. I think it is ok to leave as is in this case, but @alphan can confirm.

File: sw/device/lib/crypto/aes_gcm/aes_gcm.c [hunk 2]
```diff
@@ -399,16 +399,20 @@ aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
...
+                                     const aes_block_t *input,
+                                     aes_block_t *output) {
+  aes_error_t err = aes_encrypt_begin(key, iv);
   if (err != kAesOk) {
     return err;
   }
-  err = aes_update(output, input);
+  err = aes_update(NULL, input);
```
Comment (moidx) [inline sw/device/lib/crypto/aes_gcm/aes_gcm.c:411]: To follow the style guide: 

```c++
err = aes_update(/*dest=*/NULL, input);
```

File: sw/device/lib/crypto/drivers/aes.c [hunk 3]
```diff
@@ -67,56 +76,96 @@ aes_error_t aes_begin(aes_params_t params) {
...
+/**
+ * Configure the AES block and write the key and IV if applicable.
+ *
+ * @param key AES key.
+ * @param iv IV to use (ignored if the mode does not require an IV).
+ * @param encrypt True for encryption, false for decryption.
+ * @return result, OK or error.
+ */
+static aes_error_t aes_begin(aes_key_t key, const uint32_t iv[4],
+                             bool encrypt) {
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/aes.c:104]: Use `hardened_bool_t` instead. On the hardware side this is protected using shadow registers, but we should still consider hardening the configuration parameter in software.

File: sw/device/lib/crypto/drivers/aes.c [hunk 4]
```diff
@@ -67,56 +76,96 @@ aes_error_t aes_begin(aes_params_t params) {
...
   // ECB does not need to set an IV, so we're done early.
-  if (params.mode == kAesCipherModeEcb) {
+  if (key.mode == kAesCipherModeEcb) {
     return kAesOk;
   }
 
   uint32_t iv_offset = kBase + AES_IV_0_REG_OFFSET;
-  for (size_t i = 0; i < ARRAYSIZE(params.iv); ++i) {
-    abs_mmio_write32(iv_offset + i * sizeof(uint32_t), params.iv[i]);
+  for (size_t i = 0; i < 4; ++i) {
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/aes.c:154]: nit: Generate a constant for `4` or wrap iv in a struct to be able to use `ARRAYSIZE`

File: sw/device/lib/crypto/drivers/aes.h [hunk 5]
```diff
@@ -68,19 +68,21 @@ typedef enum aes_key_len {
...
    */
-  bool encrypt;
+  aes_cipher_mode_t mode;
 
   /**
-   * The hardware-native cipher mode to apply, if any.
+   * Whether the key is sideloaded.
    */
-  aes_cipher_mode_t mode;
+  bool sideload;
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/aes.h:85]: Lets use `hardened_bool_t` here as well.

**Comment by moidx** [comment]
Hi @jadephilipoom, please re-request review if you want me to take another look. This is just so that the PR shows up in my notification queue. Thanks!

File: sw/device/lib/crypto/drivers/aes.c [hunk 6]
```diff
@@ -67,56 +76,96 @@ aes_error_t aes_begin(aes_params_t params) {
...
+    abs_mmio_write32(iv_offset + i * sizeof(uint32_t), iv[i]);
   }
 
   return kAesOk;
 }
 
+aes_error_t aes_encrypt_begin(aes_key_t key, const uint32_t iv[4]) {
```
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/drivers/aes.c:161]: The IV is fundamentally an AES block, so I've changed all of these IV parameters to simply use the existing `aes_block_t`, which I think is a net code simplification anyway. Thanks!

**Review by moidx — APPROVED** [review]
LG. Thanks!
